### DISCUSSION TITLE
Can I save to .nii.gz by torchio easily after preprocess?

### DISCUSSION THREAD

**Discussion by Bigsealion** [TorchIO-project/torchio#731] [Q&A] [answered]
I have loaded a Nifti format file and applied some transform to this. Now, I want to save this transformed image to Nifit format to check whether my transform is right. But I don't know how to do this easily by Torchio. 

The compromise I'm using now is to save it with Nibabel package (load image by Nibabel and replace the data with transformed image), but I think it's a little clunky.

Can you tell me how to save transformed images by Torchio easily?

For example, my code is following:
```
# load data
subject_a = tio.Subject(img=tio.ScalarImage(nii_path))

# transform
rescale = tio.RescaleIntensity(out_min_max=(-1, 1), in_min_max=(0, 100))
zscore = tio.ZNormalization()

transformed = transform(subject_a)

# How to save this image to Nifit format?
# Code to save this transformed image
```

Thanks!

**Comment by romainVala** [comment]
Hi
Just try this convenient function

`transformed.img.save('/path/volume.nii')`

**Reply by Bigsealion** [reply]
Thanks! It's very helpful!

**Reply by Bigsealion** [reply]
Thanks for your reply, but I have a new question that I would like to have answered.
Thanks!

**Comment by Bigsealion** [comment, marked as answer]
When I read the [document](https://torchio.readthedocs.io/quickstart.html#hello-world), I have a new question, and the above solution did not work

The following code is from the document, and I want to know how to save the output image

```
# Transforms can be composed as in torchvision.transforms
transforms = [rescale, spatial]
transform = tio.Compose(transforms)
# SubjectsDataset is a subclass of torch.data.utils.Dataset
subjects_dataset = tio.SubjectsDataset(subjects_list, transform=transform)

# Images are processed in parallel thanks to a PyTorch DataLoader
training_loader = DataLoader(subjects_dataset, batch_size=4, num_workers=4)

# Training epoch
for subjects_batch in training_loader:
    inputs = subjects_batch['t1'][tio.DATA]
    target = subjects_batch['label'][tio.DATA]

# how to save the inputs and target to .nii.gz by Torchio?
```
Thanks!

**Reply by romainVala** [reply]
just the same command
subjects_batch.t1.save('t1.nii')
subjects_batch.label.save('label.nii')

**Reply by fepegar** [reply]
@romainVala that wouldn't work because `subjects_batch` is a `dict` generated by the data loader. 

@Bigsealion you could instantiate the images manually using `ScalarImage` or `LabelMap`, or you could try `get_subjects_from_batch`:

```python
In [3]: tio.utils.get_subjects_from_batch?
Signature: tio.utils.get_subjects_from_batch(batch: Dict) -> List
Docstring:
Get list of subjects from collated batch.

Args:
    batch: Dictionary generated by a :class:`torch.utils.data.DataLoader`
    extracting data from a :class:`torchio.SubjectsDataset`.
File:      ~/git/torchio/torchio/utils.py
Type:      function
```

**Reply by Bigsealion** [reply]
Thanks! 
get_subjects_from_batch is a useful function for checking these out files!

And, I still have two small questions.

First: How to instantiate the images manually using ScalarImage? following code
`new_subj=tio.ScalarImage(subjects_batch)` or `new_subj=tio.ScalarImage(subjects_batch['t1'])` is wrong

Second: How to run this command `tio.utils.get_subjects_from_batch?`? I have run this command in the Python console by only getting the following error:
```
  File "<input>", line 1
    tio.utils.get_subjects_from_batch?
                                     ^
SyntaxError: invalid syntax
```

Thanks for your kind reply!

**Reply by fepegar** [reply]
> First: How to instantiate the images manually using ScalarImage? following code
new_subj=tio.ScalarImage(subjects_batch) or new_subj=tio.ScalarImage(subjects_batch['t1']) is wrong

That doesn't work because `subjects_batch['t1']` is a `dict`. You can use the `tensor` and `affine` kwargs of `ScalarImage` to instantiate it. There are many examples in the quick start page and in the tutorials. The tensor data you want is, for example, in `subjects_batch['t1'][tio.DATA][0]`.

> Second: How to run this command tio.utils.get_subjects_from_batch?? I have run this command in the Python console

I was using IPython (or Jupyter). In bare Python, you can run `help(tio.utils.get_subjects_from_batch)`. But I recommend IPython over Python for experimentation.

**Reply by Bigsealion** [reply]
Thanks for your reply!
Now, I know the meaning of 'manually', and the command `help()` is helpful!
